### PR TITLE
Feat/multi index

### PR DESCRIPTION
**Pull request by PLNech**
Provides multi index capabilities.
- [x] Instantiating InstantSearch with an activity using several Fragments
- [x] Creating several searchers
- [x] Filtering DataBinding's bindings based on the widget group
- [x] Discriminating events depending on the Searcher that triggered them
- [x] Registering all widgets in an Activity, discriminating between variants
- [x] Having a SearchView call several searchers, not only the last one registered

Usage as seen in [movies example app](https://github.com/algolia/instantsearch-android-examples/tree/feat/moviesDemo#movies-application):
```java
// Instantiate your Searchers
searcherMovies = Searcher.create(ALGOLIA_APP_ID, ALGOLIA_API_KEY, ALGOLIA_INDEX_MOVIES);
searcherActors = Searcher.create(ALGOLIA_APP_ID, ALGOLIA_API_KEY, ALGOLIA_INDEX_ACTORS);
// Create a SearchBoxViewModel with your SearchBox
searchBoxViewModel = new SearchBoxViewModel(this.findViewById(R.id.include_searchbox););
// Link it to your search widgets
new InstantSearch(this, searcherMovies).registerSearchView(this, searchBoxViewModel);
new InstantSearch(this, searcherActors).registerSearchView(this, searchBoxViewModel);
```
(Will write a guide once the PR is approved 😉)

### CODE REVIEW

File: instantsearch/src/main/java/com/algolia/instantsearch/helpers/InstantSearch.java [hunk 1]
```diff
@@ -398,9 +465,9 @@ private void prepareWidget(@Nullable View rootView, View widget, @Nullable List<
                 // Link hits to activity's empty view //TODO: Remove rootView parameter if getRootView always works
                 ((Hits) widget).setEmptyView(getEmptyView(widget.getRootView()));
 
-                itemLayoutId = ((Hits) widget).getLayoutId();
+                int itemLayoutId = ((Hits) widget).getLayoutId();
 
-                if (itemLayoutId == -42) {
+                if (itemLayoutId == 0) {
```
Comment (robertmogos) [inline instantsearch/src/main/java/com/algolia/instantsearch/helpers/InstantSearch.java]: Why the change ?

File: instantsearch/src/main/java/com/algolia/instantsearch/helpers/Searcher.java [hunk 2]
```diff
@@ -103,55 +106,116 @@
...
+     * @param index   an Index initialized and eventually configured.
      * @return the new instance.
      */
     @SuppressWarnings({"WeakerAccess", "unused"}) // For library users
-    public static Searcher create(@NonNull final String appId, @NonNull final String apiKey, @NonNull final String indexName) {
-        return create(new Client(appId, apiKey).getIndex(indexName));
+    public static Searcher create(@NonNull final Index index, @NonNull String variant) {
+        boolean shouldReplace = false;
+        Searcher instance = instances.get(variant);
+        if (instance != null) {
```
Comment (robertmogos) [inline instantsearch/src/main/java/com/algolia/instantsearch/helpers/Searcher.java]: Instead of 2 ifs:

if (instance == null || instance.getIndex() != index) { .. replace logic }

Less code and no temp variable

File: instantsearch/src/main/java/com/algolia/instantsearch/helpers/Searcher.java [hunk 3]
```diff
@@ -929,6 +992,17 @@ public void requestCompleted(JSONObject content, AlgoliaException error) {
         });
     }
 
+    @Override public String toString() {
+        String key = null;
```
Comment (robertmogos) [inline instantsearch/src/main/java/com/algolia/instantsearch/helpers/Searcher.java]: For cases like this we could use BiMap in the future to avoid writing redundant code. https://google.github.io/guava/releases/19.0/api/docs/com/google/common/collect/BiMap.html

File: instantsearch/src/main/java/com/algolia/instantsearch/model/SearchBoxViewModel.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+                return true;
+            }
+
+            @Override
+            public boolean onQueryTextChange(String newText) {
+                EventBus.getDefault().post(new QueryTextChangeEvent(newText, searchViewFacade.getSearchView()));
+                final String query = searchViewFacade.getQuery().toString();
+
+                for (InstantSearch instantSearch : listeners) {
+                    if (newText.length() != 0 || instantSearch.hasSearchOnEmptyString()) {
```
Comment (robertmogos) [inline instantsearch/src/main/java/com/algolia/instantsearch/model/SearchBoxViewModel.java]: use .isEmpty instead of length. You have better runtime complexity

File: instantsearch/src/main/java/com/algolia/instantsearch/ui/databinding/BindingHelper.java [hunk 5]
```diff
@@ -1,128 +1,193 @@
...
+     * @return the previous variant for this view, if any.
+     */
+    @Nullable public static String setVariantForView(@NonNull View view, @Nullable String variant) {
+        String previousVariant = null;
+        String previousAttribute = null;
+        for (Map.Entry<String, HashMap<Integer, String>> entry : bindings.entrySet()) {
+            if (entry.getValue().containsKey(view.getId())) {
+                previousVariant = entry.getKey();
+                previousAttribute = entry.getValue().get(view.getId());
+                bindings.remove(previousVariant);
```
Comment (robertmogos) [inline instantsearch/src/main/java/com/algolia/instantsearch/ui/databinding/BindingHelper.java]: Is this safe? Modifying the collection while iterating ? I would do a two step process. First find then remove

File: instantsearch/src/main/java/com/algolia/instantsearch/ui/views/Hits.java [hunk 6]
```diff
@@ -551,17 +555,37 @@ private Activity getActivity(View view) {
...
+                String indexVariant = defaultValue;
+
+                // Get the index and variant for this layout
+                final List<View> views = LayoutViews.findAny((ViewGroup) itemView);
+                for (View view : views) {
+                    if (view instanceof AlgoliaHitView) {
+                        continue;
+                    }
+                    String viewIndexVariant = BindingHelper.getVariantForView(view);
+                    final boolean isDifferentVariant = !(viewIndexVariant == null ? indexVariant == null : viewIndexVariant.equals(indexVariant));
```
Comment (robertmogos) [inline instantsearch/src/main/java/com/algolia/instantsearch/ui/views/Hits.java]: rename isSameVariant and remove the ! before. Add it to the next if. Feels easier to read

File: instantsearch/src/main/java/com/algolia/instantsearch/helpers/InstantSearch.java [hunk 7]
```diff
@@ -85,40 +86,96 @@ public InstantSearch(@NonNull final Activity activity, @NonNull final Searcher s
...
     }
 
     /**
-     * Registers the Search Widget of an Activity's Menu to trigger search requests on text change.
+     * Triggers a new search with the given text.
+     *
+     * @param query the text to search for.
+     */
+    public void search(String query) {
+        searcher.setQuery(searcher.getQuery().setQuery(query)).search();
```
Comment (spinach) [inline instantsearch/src/main/java/com/algolia/instantsearch/helpers/InstantSearch.java]: A lot of things are happening here, If you can write separate statements for that it would be great!

File: instantsearch/src/main/java/com/algolia/instantsearch/helpers/InstantSearch.java [hunk 8]
```diff
@@ -245,20 +323,6 @@ public void registerFilters(List<AlgoliaFilter> filters) {
...
     }
 
-    /**
-     * Gets the {@link android.support.annotation.IdRes IdRes} of the item layout registered by a Hits widget.
-     * This method should only be used by the library's internals.
-     *
-     * @return the registered item layout id, if any.
-     * @throws IllegalStateException if no item layout was registered.
-     */
-    public static int getItemLayoutId() {
```
Comment (spinach) [inline instantsearch/src/main/java/com/algolia/instantsearch/helpers/InstantSearch.java]: make sure you're not breaking BC here

File: instantsearch/src/main/java/com/algolia/instantsearch/helpers/Searcher.java [hunk 9]
```diff
@@ -103,55 +106,116 @@
...
         final LibraryVersion version = new LibraryVersion("InstantSearch Android", String.valueOf(BuildConfig.VERSION_NAME));
         if (!Arrays.asList(client.getUserAgents()).contains(version)) {
             client.addUserAgent(version);
         }
     }
 
+    /**
+     * Destroys all Searcher instances.
+     */
+    public static void destroyAll() {
```
Comment (spinach) [inline instantsearch/src/main/java/com/algolia/instantsearch/helpers/Searcher.java]: nit: suggestion: `clearAll`

File: instantsearch/src/main/java/com/algolia/instantsearch/helpers/InstantSearch.java (hunk 1 above)
Comment (PLNech) [reply instantsearch/src/main/java/com/algolia/instantsearch/helpers/InstantSearch.java]: 0 is not a valid`ResId` value, hence no need for a magic number

File: instantsearch/src/main/java/com/algolia/instantsearch/helpers/Searcher.java (hunk 3 above)
Comment (PLNech) [reply instantsearch/src/main/java/com/algolia/instantsearch/helpers/Searcher.java]: 👍, adding a todo for later refactoring then!

File: instantsearch/src/main/java/com/algolia/instantsearch/model/SearchBoxViewModel.java (hunk 4 above)
Comment (PLNech) [reply instantsearch/src/main/java/com/algolia/instantsearch/model/SearchBoxViewModel.java]: Spot on! Applied whenever possible.

File: instantsearch/src/main/java/com/algolia/instantsearch/ui/databinding/BindingHelper.java (hunk 5 above)
Comment (PLNech) [reply instantsearch/src/main/java/com/algolia/instantsearch/ui/databinding/BindingHelper.java]: I thought iterating on the entrySet was safe, but you're right:

File: instantsearch/src/main/java/com/algolia/instantsearch/ui/views/Hits.java (hunk 6 above)
Comment (PLNech) [reply instantsearch/src/main/java/com/algolia/instantsearch/ui/views/Hits.java]: Fair point.

File: instantsearch/src/main/java/com/algolia/instantsearch/helpers/InstantSearch.java (hunk 8 above)
Comment (PLNech) [reply instantsearch/src/main/java/com/algolia/instantsearch/helpers/InstantSearch.java]: Fair concern! However, as there was no reason to use it from a user's perspective, and as I warned in the docs `This method should only be used by the library's internals`, I'm confident this is a harmless change :)

File: instantsearch/src/main/java/com/algolia/instantsearch/helpers/Searcher.java (hunk 9 above)
Comment (PLNech) [reply instantsearch/src/main/java/com/algolia/instantsearch/helpers/Searcher.java]: Open for debate: I called it `destroyAll` as it's meant to be used in the `Activity`/`Fragment`'s **`onDestroy`** method.

Comment (robertmogos) [reply instantsearch/src/main/java/com/algolia/instantsearch/helpers/Searcher.java]: @PLNech @spinach destroy is the Android specific name. We should go with that
